### DISCUSSION TITLE
Bashing competitors without facts is counterintuitive

### DISCUSSION THREAD

**Discussion by uuf6429** [Codeception/Codeception#6078] [General]
Disclaimer: I've used Behat for a couple of projects extensively, so I know its upsides and downsides.

For a new project, I decided to reevaluate codeception as a solution and, long story short, I ended up on this page:
https://codeception.com/12-20-2012/not-bdd.html

In short, I decided to NOT use Codeception after reading that article due to how it is written. Here's a few points why:
- language/grammar is quite poor, from the start it gives a bad impression of the writer
- it exaggerates, invents problems or simply misdirects the reader on several points:
  - if someone changes the responses of the system under test, the tests have to be changed - doesn't matter if it's Codeception or Behat
  - BDD 'features' are made for QA engineers (yes, they typically would know their tool - selenium/java/php/etc) - if your manager wants to write tests, it is his problem - same if he wants to write Codeception tests
  - ^that also answers why it is written in plain English. BTW, Codeception is also in English (or tries to) ie. `$I->click('Order');` instead of `$page->dispatch(new MouseEvent(MouseEvent::LEFT_BUTTON));` 😉 
  - The "marketing trap" is a stupid point - users/devs do not need to write tests twice and in fact encourages to be more DRY than Codeception (by creating reusable steps, whereas Codeception, according to that article, encourages to write big scripts testing everything - which as far as testing goes, is a bad idea)
  - KISS is also a stupid point - you can argue that dependency injection, containers, autoloading are all not KISS
- nice to have a disclaimer, but maybe it's better to be on top of the page - by the time I got there I was mostly incredulous to be reading this content directly on codeception.com
- by the way, Behat can also be used to write non-BDD tests (similar to Codeception), so would have been nice to see this compared
- on a final point, the article conveniently forgot to mention anything about Codeception testing the non-public interface (ie, using css selectors) - which is generally known as a bad idea in these sort of tests, so I was hoping/expecting to read about why it doesn't apply to Codeception.

In the end, I was hoping to find some information on why I should use Codeception, instead of Behat (possible trimmed down) or even plain PHPUnit with Mink(for example) and instead I got a poorly-worded, misdirected bash of a competitor. 🤷‍♂️

**Comment by DavertMik** [comment]
Ok, you read a blogpost which is 8 years old. If you follow the modern trends and technologies you may notice that LOT of things has changed since that time. So I don't want to discuss something that was written 8 years ago. I don't have a time machine. Maybe you do.

If you like Behat - use it
If you like Codeception - use it.

That's simple.

**Reply by uuf6429** [reply]
> Ok, you read a blogpost which is 8 years old.

It was the top result in google for a comparison between behat and codeception. Besides, the problem is not with it being outdated, but by the content. It is opinion-based and not backed by any facts, hosted(=endorsed by) on codeception.com. I know you guys are devs, but there's this not so new thing called public relations and marketing.

> So I don't want to discuss something that was written 8 years ago.

You don't have to. You can change the content, label it properly or take it down. Or even do nothing. It's all up to you.

> If you like Behat - use it
> If you like Codeception - use it.

Yeah. Great way of getting new users.

Anyhow, for the record, I ended up going with `symfony/panther`.

**Reply by DavertMik** [reply]
> Anyhow, for the record, I ended up going with symfony/panther.

Good to know. Keep us updated!

P.S> When I google technical information I always set 1 year filter. That really helps

**Reply by uuf6429** [reply]
> Good to know. Keep us updated!

In about 4 hours I've installed panther+driver (this took about 10 mins, mostly spent reading the panther doc), set up the first feature test (installing a plugin in wordpress), did some finishing touches (taking screenshot on error, assertions etc - panther by itself is not specific for testing) and set up a github workflow (I'm completely new to github workflows, that took most of the time).
All in all, quite happy with the result.

Don't know about codeception, but I guess that after 4 hours I'd still be fighting with the behat config. 😆

**Comment by Naktibalda** [comment]
@DavertMik some warning sign that it is an opinion blog post and that it is horribly outdated would be nice though.

Even the answer to second question (Why can’t I write Codeception tests in human language like Behat does) has changed.

**Reply by DavertMik** [reply]
Indeed, Codeception got BDD support since that time :smile: 
I don't know, there are still valid points there, to my experience from the last year BDD can have very bad consequences for test automation on a big project.

So if the person who writes gherkin scenarios is not the person who automates them
In this case, that person has no idea of the context of steps, their dependencies, and preconditions. This leads to a very bad scenarios and a complete disaster to test automation engineer. Instead of writing really working tests they spend their work hours in complete refactoring of steps for each scenario change.

If the person who writes scenarios and implements test automation is the same person - they have pretty much the same problem (constant steps refactoring), but also this doesn't give any value on top. It makes sense to have BDD scenarios as public documentation. But still with the cost of extra coding, refactoring, etc. 

People tend to misuse BDD, that's still a valid thing. People don't take tradeoffs of BDD into account implementing it.
So the points in the post are still valid. I don't have time to re-read and update the post. As lots of other posts written in that blog.

**Reply by uuf6429** [reply]
> I don't know, there are still valid points there, to my experience from the last year BDD can have very bad consequences for test automation on a big project.

Such as? I said it before, this whole crusade against BDD without any factual background gives the impression that it's just a subjective rant without much knowledge on how it really works.

When writing gherkin steps, a competent IDE would suggest / autocomplete existing steps.
Steps can be parameterised to be reusable - if done properly, you don't need to change the code.
Frankly, I can see the big disaster coming out of non-BDD tests because of developers writing 200+loc test methods with huge setups and multiple assertions in the same method.
The main value is splitting business logic from test logic, not documentation. BDD by default splits this logic.

Yes, there are tradeoffs and annoyances, but unfortunately, I haven't seen them mentioned or compared against traditional test systems / codeception in that blog entry.. all I'm saying is that if the point of it was to convince me to switch to codeception, then it failed badly. If that wasn't the point, then you tell me.

**Reply by DavertMik** [reply]
Ok, you say that I don't give arguments but...

> I can see the big disaster coming out of non-BDD tests because of developers writing 200+loc test methods with huge setups and multiple assertions in the same method.

How is this even related to BDD?

> The main value is splitting business logic from test logic, not documentation. BDD by default splits this logic.

This also has no relation to BDD. 

You can develop small and meaningful tests without using BDD principles such us:

* test-first approach
* writing Gherkin scenarios
* trying to make a test to be a specification 

Just
* focus on things that matter (business logic, complex algorithms, end2end scenarios)
* write atomic independent tests
* automate all the things in tests

Since 2012 we have a much better guide on BDD and its limits and how BDD can be combined with classical tests to resolve those issues https://codeception.com/docs/07-BDD#Tests-vs-Features

I don't know why did you put all *good code practices* as a consequence of BDD. No, they are not even related. BDD is  more about management and inter-team communication in a project. Did you hear "3-Amigo" term? Automated tests are a minor part of BDD. And following BDD principles can make harden the development of those tests. That's my point.

This is why the post may sound counterintuitive to you. And this is why BDD can be harmful. People tend to think that BDD will solve all their issues - with tests, with clean code, with team communication. No. BDD is a set of principles that can either improve or harm test development, with its pros and cons. Anyway, unless you are following 3-Amigo principle - you are not doing real BDD.

Problem is not in BDD itself. But with how people think of it.

**Reply by uuf6429** [reply]
You've got a point that BDD, from _a general perspective_, isn't related to code at all, but more of a high-level development principle.
However, here we're talking about _BDD within Behat_. The article's introduction starts with **"behat vs codeception"**, not **"bdd vs codeception / something else"**.

> > I can see the big disaster coming out of non-BDD tests because of developers writing 200+loc test methods with huge setups and multiple assertions in the same method.
>
> How is this even related to BDD?

When writing the user stories, you're forced to keep out implementation details. Again, within the context of Behat.

> > The main value is splitting business logic from test logic, not documentation. BDD by default splits this logic.
>
> This also has no relation to BDD.


If we agree that behaviour is basically business logic, then as BDD goes, it means the behaviour/business-logic is written first - and then connected to the test system - therefore: clear separation between business and test logic.
I can also quote that article on it actually:
> The Behavior Driven Development concept requires that the User Stories to be written, and then executed as a test.

> I don't know why did you put all _good code practices_ as a consequence of BDD. No, they are not even related.

I could ask that question as well. The article claims that using Behat/BDD is bad for DRY/KISS etc (when in fact it encourages and forces code reuse). Why?

Behat (or mink, actually), by default, starts with a restrictive set of steps that should immediately give the idea that using selectors for everything in your tests is a bad idea. Codeception (within that article) and similar systems (like panther) start with a different ("more flexible") approach, without saying that it is a bad practice (quite the contrary, the article claims that using selectors everywhere is a good thing). 

> This is why the post may sound counterintuitive to you. And this is why BDD can be harmful. People tend to think that BDD will solve all their issues - with tests, with clean code, with team communication. No. BDD is a way to formalize communication. The quality of your automated tests is absolutely up to you.

Agreed, from an organisational point of view, anything could screw up. However, the morality of that point is: _it's not about the tool, but the usage_. So you should now see why I wasn't expecting that point on a post _comparing two tools_.

**Reply by DavertMik** [reply]
> The article's introduction starts with "behat vs codeception", not "bdd vs codeception / something else".

Yes. In 2012 there was a huge public attack from Behat team on Codeception. And this post was my attempt to respond to it. Sure, the goal of this post was what it said - propose alternative to Behat and show weak points of Behat's approach.

> If we agree that behaviour is basically business logic, then as BDD goes, it means the behaviour/business-logic is written first - and then connected to the test system - therefore: clear separation between business and test logic.
I can also quote that article on it actually:

Behavior is not really a business logic. Business logic is a decision unit inside an application while application behavior it what user sees. So when user clicks on site it takes down through all sub-systems to make that behavior works. But a business logic is only responsible to data changes within this action. 

For instance, I can write my behavior test on any of these levels:

* End 2 End tests (via browser)
* API test (this API is used by browser)
* Controller test
* Service test

The test layer that has the least dependencies on other sub-modules is testing business logic. 
Test logic goes from the layer it is executed. And in a good system, there should be a test for each of the specific layers covering different behavior cases. 

Let's take REST API cases - should we write them in BDD manner as Gherkin scenarios? No, REST API is a technical interface, so usage of human-like language is not really needed. Should we test REST API before providing it to our React/Vue app? Sure! Where is BDD here? I don't know.

> Behat (or mink, actually), by default, starts with a restrictive set of steps that should immediately give the idea that using selectors for everything in your tests is a bad idea. Codeception (within that article) and similar systems (like panther) start with a different ("more flexible") approach, without saying that it is a bad practice (quite the contrary, the article claims that using selectors everywhere is a good thing).

The idea of coupling Behat with Mink was also later taken as anti-pattern by Behat team (sorry, can't find their tweet from 2012). UI steps bound to user scenarios does not tell anything of application behavior, Gherkin scenarios transform into interface guides. If we change the button's position on the page the business scenario is not relevant anymore? BDD tests in Behat should be kept around business logic even when defining how to work with an application's interface. 

The approach of Codeception is quite different. We say that our tests are indeed user interface tests. And use any selectors you need to pass the scenario. We are focusing on tests. Cover your application with the required set of tests you need. This is absolutely ok.

In BDD scenarios you use human-like language to define business scenarios 
In Codeception you write browser tests.

What is antipattern in Behat+Mink is a pattern in Codeception. Just because of the different purposes of these tools.

> Agreed, from an organisational point of view, anything could screw up. However, the morality of that point is: it's not about the tool, but the usage. So you should now see why I wasn't expecting that point on a post comparing two tools.

Sorry, for my rude start. I understand this. But this post has its own context. And unfortunately, I can't reveal all the context it was created under. Codeception was attacked by Behat author who was more popular in PHP community than me. So maybe that what triggered me first. Yes, in that year this post was about Codeception vs Behat. 

Nowadays, things have changed and Codeception got its place under the sun. 
The reason the post is still there - it is still getting the attention of new users and gives conversion. So I prefer to keep it as it is.
